galley/pkg/config/analysis/analyzers/annotations/annotations.go

```diff
@@ -0,0 +1,219 @@
+	resourceTypeNames = map[annotation.ResourceTypes]string{
+		annotation.Ingress:      "Ingress",
+		annotation.Pod:          "Pod",
+		annotation.Service:      "Service",
+		annotation.ServiceEntry: "ServiceEntry",
+	}
+)
+
+// Metadata implements analyzer.Analyzer
+func (*K8sAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "annotations.K8sAnalyzer",
+		Inputs: collection.Names{
+			metadata.K8SCoreV1Endpoints,
+			metadata.K8SCoreV1Namespaces,
+			metadata.K8SCoreV1Services,
+			metadata.K8SCoreV1Pods,
+			metadata.K8SExtensionsV1Beta1Ingresses,
+			metadata.K8SCoreV1Nodes,
+		},
+	}
+}
+
+// Analyze implements analysis.Analyzer
+func (fa *K8sAnalyzer) Analyze(ctx analysis.Context) {
+	ctx.ForEach(metadata.K8SCoreV1Endpoints, func(r *resource.Entry) bool {
+		fa.allowAnnotations(r, ctx, "Endpoints", metadata.K8SCoreV1Endpoints)
+		return true
+	})
+	ctx.ForEach(metadata.K8SCoreV1Namespaces, func(r *resource.Entry) bool {
+		fa.allowAnnotations(r, ctx, "Namespace", metadata.K8SCoreV1Namespaces)
+		return true
+	})
+	ctx.ForEach(metadata.K8SCoreV1Nodes, func(r *resource.Entry) bool {
+		fa.allowAnnotations(r, ctx, "Node", metadata.K8SCoreV1Nodes)
+		return true
+	})
+	ctx.ForEach(metadata.K8SCoreV1Pods, func(r *resource.Entry) bool {
+		fa.allowAnnotations(r, ctx, "Pod", metadata.K8SCoreV1Pods)
+		return true
+	})
+	ctx.ForEach(metadata.K8SCoreV1Services, func(r *resource.Entry) bool {
+		fa.allowAnnotations(r, ctx, "Service", metadata.K8SCoreV1Services)
+		return true
+	})
+	ctx.ForEach(metadata.K8SExtensionsV1Beta1Ingresses, func(r *resource.Entry) bool {
+		fa.allowAnnotations(r, ctx, "Ingress", metadata.K8SExtensionsV1Beta1Ingresses)
+		return true
+	})
+}
+
+func (*K8sAnalyzer) allowAnnotations(r *resource.Entry, ctx analysis.Context, kind string, collectionType collection.Name) {
+	if len(r.Metadata.Annotations) == 0 {
+		return
+	}
+
+	// It is fine if the annotation is kubectl.kubernetes.io/last-applied-configuration.
+	for ann := range r.Metadata.Annotations {
+		if istioAnnotation(ann) {
+
+			annotationDef := lookupAnnotation(ann)
+			if annotationDef == nil {
+				ctx.Report(collectionType,
+					msg.NewUnknownAnnotation(r, ann))
+				continue
+			}
+
+			attachesTo := resourceTypesAsStrings(annotationDef.Resources)
+			if !contains(attachesTo, kind) {
+				ctx.Report(collectionType,
+					msg.NewMisplacedAnnotation(r, ann, strings.Join(attachesTo, ", ")))
+				continue
+			}
+
+			// TODO: Check annotation.Deprecated.  Not implemented because no
+			// deprecations in the table have yet been deprecated!
+
+			// Note: currently we don't validate the value of the annotation,
+			// just key and placement.
+			// There is annotation value validation (for pod annotations) in the web hook at
+			// istio.io/istio/pkg/kube/inject/inject.go.  Rather than refactoring
+			// that code I intend to bring all of the web hook checks into this framework
+			// and checking here will be redundant.
```

Comment:
Link to an issue here?

Reply:
I thought you would do this under https://github.com/istio/istio/issues/16777 .  If you do not want to write the analyzers I will take a crack at it.

Reply:
I did https://github.com/istio/istio/pull/17770, which covers schema validation. 

It doesn't look like validateAnnotations (in inject.go) gets called anywhere by the validation webhook or the current `istioctl validate`. The issue you referenced doesn't cover this, so we should probably create a new one that does.